crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,275 @@
+multiple `str` instances equal to `"foo"` can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+An example of a singleton type that is not a single-value type
+would be `Literal[Ham.A]` for the following `Ham` enum:
+because the `Ham` class overrides `__eq__`, we can no longer say for sure
+that the `Ham` type can be safely narrowed using inequality as in the above example
+(but we *can* still narrow the `Ham` type using identity, as with all singleton types):
+
+```py
+from enum import Enum, auto
+
+class Ham(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+    def __eq__(self, other):
+        return True
+```
+
+## Final types
+
+A final type in Python is a [nominal type](https://docs.python.org/3/library/typing.html#nominal-vs-structural-subtyping)
+defined as the instances of a certain runtime class that it is known cannot be subclassed.
+Most final types in Python are associated with class definitions that are decorated with
+[`@typing.final`](https://docs.python.org/3/library/typing.html#typing.final).
+This is a special decorator which indicates to the type checker that the tool
+should assume the class cannot be subclassed, and that the tool should emit an error
+if it detects an attempt to do so.
+
+There are fewer ways in which final types can be subtyped when compared to most other types.
+For any two nominal types `X` and `Y`, if `X` is `@final` and `X` is not a subtype of `Y`,
+the intersection `X & Y` is equivalent to `Never`, the empty set.
+However, a final type in Python still has subtypes, just no *nominal* subtypes.
+For the following two types `X` and `Truthy`, the intersection `X & Truthy` is non-empty,
+since `Truthy` is a structural type:
+
+```py
+from typing import final, Protocol, Literal
+
+class Truthy(Protocol):
+    """Any object `x` inhabits this type if `bool(x)` always returns `True`"""
+
+    def __bool__(self) -> Literal[True]: ...
+
+@final
+class X[T: bool]:
+    def __init__(self, bool_val: T):
+        self.bool_val = bool_val
+
+    def __bool__(self) -> T:
+        return self.bool_val
+```
```

Comment:
I think this section is a bit debatable.

If we are talking about a conceptual `Truthy` type consisting of "all objects whose `__bool__` actually returns `True`", that might work for this example, but it is probably not a good example to include since it's not clear if we even want to define such a type, given that individual objects can easily change their `__bool__` behavior over time.

If we are talking about the Protocol type shown here, then given a final type X, `X & Truthy` will always be either `X` or `Never`, it will never be a proper subtype of `X`, because `X.__bool__` has a known signature, and that signature is either a subtype of `(self) -> Literal[True]` or it is not.

A better example of proper subtypes of final types might be a final generic type with a covariant type parameter? In that case `X[B] <: X[A]` if `B <: A`, even though `X` is final.

At a higher level, though, it remains unclear to me why it is important to discuss whether a final type has any proper subtypes here?

Reply:
I find it pretty useful to remember that runtime subclassing and static subtyping are linked but not the same thing. Just because something is `@final` does not mean it cannot be subtyped, only that it cannot be subclassed; all types except singleton types can be subtyped.